### PR TITLE
many: generalize assertstate.Batch to asserts.Batch, have assertstate.AddBatch

### PR DESCRIPTION
**Pull request by pedronis**
This generalizes assertstate.Batch to asserts.Batch which grows a bit more functionality. To achieve the same result, assertstate.AddBatch subsuming assertstate.Batch.Precheck/Commit is introduced.

doFetch in assertstate is also switched to use asserts.Batch which can be built by fetching for convenience.

The fallback code for dealing with receiving unsupported format assertions, which should not happen except for bugs, is reorganized.

All now unused bits in assertstate are dropped.

### CODE REVIEW

File: asserts/batch.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+			}
+		}
+		return err
+	}
+	b.added = append(b.added, a)
+	return nil
+}
+
+// AddStream adds a stream of assertions to the batch.
+// Returns references to to the assertions effectively added.
```
Comment (bboozzoo) [inline asserts/batch.go]: `s/to to/to/`

File: asserts/batch.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		return nil, nil
+	}
+	refs := make([]*Ref, len(added))
+	for i, a := range added {
+		refs[i] = a.Ref()
+	}
+	return refs, nil
+}
+
+// Fetch adds to the batch via a driven by the fetching function
```
Comment (bboozzoo) [inline asserts/batch.go]: I'd reword this to:
```
Fetch adds to the batch via the internal Fetcher built with trustedDB and retrieve.
The fetcher is driven the by provided fetching function.
```

File: asserts/batch_test.go [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+		headers[h] = v
+	}
+	decl, err := s.storeSigning.Sign(asserts.SnapDeclarationType, headers, nil, "")
+	c.Assert(err, IsNil)
+	err = s.storeSigning.Add(decl)
+	c.Assert(err, IsNil)
+	return decl.(*asserts.SnapDeclaration)
+}
+
+func (s *batchSuite) TestAddStream(c *C) {
```
Comment (bboozzoo) [inline asserts/batch_test.go]: For completeness, a test with an empty stream could be useful too.

File: asserts/batch.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+// Precheck pre-checks whether adding the batch of assertions to the
+// given assertion database should fully succeed.
+func (b *Batch) Precheck(db *Database) error {
+	db = db.WithStackedBackstore(NewMemoryBackstore())
+	return b.commitTo(db)
+}
+
+// CommitTo adds the batch of assertions to the given assertion database.
```
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/batch.go]: Thanks for the input on this, there's been quite a bit of back and forth. That's what I did in assertstate, I can probably port that approach back here. It's just a bigger change and I'll have to expor_test Precheck to have better testing control.

File: asserts/batch.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		return err
+	}
+	b.added = append(b.added, a)
+	return nil
+}
+
+// AddStream adds a stream of assertions to the batch.
+// Returns references to the assertions effectively added.
+func (b *Batch) AddStream(r io.Reader) ([]*Ref, error) {
+	b.inPrereqOrder = false
```
Comment (zyga) [inline asserts/batch.go]: This assignment could move a little bit down, after len(added) check.

File: asserts/batch.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+func (b *Batch) prereqSort(db *Database) error {
+	if b.inPrereqOrder {
+		// nothing to do
+		return nil
+	}
+
+	// put in prereq order using a fetcher
+	inPrereqOrder := make([]Assertion, 0, len(b.added))
```
Comment (zyga) [inline asserts/batch.go]: This variable is named the same as the boolean flag which is mildly confusing. I would suggest giving it a new name, e.g. just `sorted`

**Comment by pedronis** [comment]
merging given it's green, but will do a follow up

File: asserts/batch_test.go [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+	// snap-declaration was added anyway
+	_, err = s.db.Find(asserts.SnapDeclarationType, map[string]string{
+		"series":  "16",
+		"snap-id": "foo-id",
+	})
+	c.Assert(err, IsNil)
+}
+
+func (s *batchSuite) TestCommitMissing(c *C) {
+	// store key already present
```
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/batch_test.go]: just consistency with the other tests setup, also makes the test rely less on the internal order in which Fetcher does things
